examples/mpt/main.cpp

```diff
@@ -0,0 +1,737 @@
+               name.data(), ne[0], ne[1], ggml_type_name(ggml_type(ttype)),
+               ggml_nbytes(tensor) / 1024.0 / 1024.0, ggml_nbytes(tensor));
+      }
+
+      const size_t bpe = ggml_type_size(ggml_type(ttype));
+
+      if ((nelements * bpe) / ggml_blck_size(tensor->type) !=
+          ggml_nbytes(tensor)) {
+        fprintf(stderr,
+                "%s: tensor '%s' has wrong size in model file: got %zu, "
+                "expected %zu\n",
+                __func__, name.data(), ggml_nbytes(tensor), nelements * bpe);
+        return false;
+      }
+
+      fin.read(reinterpret_cast<char *>(tensor->data), ggml_nbytes(tensor));
+
+      total_size += ggml_nbytes(tensor);
+      if (++n_tensors % 8 == 0) {
+        printf(".");
+        fflush(stdout);
+      }
+    }
+
+    printf(" done\n");
+
+    printf("%s: model size = %8.2f MB / num tensors = %d\n", __func__,
+           total_size / 1024.0 / 1024.0, n_tensors);
+  }
+
+  fin.close();
+
+  return true;
+}
+
+// evaluate the transformer
+//
+//   - model:     the model
+//   - n_threads: number of threads to use
+//   - n_past:    the context size so far
+//   - embd_inp:  the embeddings of the tokens in the context
+//   - embd_w:    the predicted logits for the next token
+//
+bool mpt_eval(const mpt_model &model, const int n_threads, const int n_past,
+              const std::vector<gpt_vocab::id> &embd_inp,
+              std::vector<float> &embd_w, size_t &mem_per_token) {
+  const int N = embd_inp.size();
+
+  const auto &hparams = model.hparams;
+
+  const int n_embd = hparams.d_model;
+  const int n_layer = hparams.n_layers;
+  const int n_ctx = hparams.max_seq_len;
+  const int n_head = hparams.n_heads;
+  const int n_vocab = hparams.n_vocab;
+
+  static size_t buf_size = 256u * 1024 * 1024;
+  static void *buf = malloc(buf_size);
+
+  if (mem_per_token > 0 && mem_per_token * N > buf_size) {
+    const size_t buf_size_new =
+        1.1 *
+        (mem_per_token * N); // add 10% to account for ggml object overhead
+    // printf("\n%s: reallocating buffer from %zu to %zu bytes\n", __func__,
+    // buf_size, buf_size_new);
+
+    // reallocate
+    buf_size = buf_size_new;
+    buf = realloc(buf, buf_size);
+    if (buf == nullptr) {
+      fprintf(stderr, "%s: failed to allocate %zu bytes\n", __func__, buf_size);
+      return false;
+    }
+  }
+
+  struct ggml_init_params params = {
+      .mem_size = buf_size,
+      .mem_buffer = buf,
+      .no_alloc = false,
+  };
+
+  struct ggml_context *ctx0 = ggml_init(params);
+  struct ggml_cgraph gf = {.n_threads = n_threads};
+
+  struct ggml_tensor *embd = ggml_new_tensor_1d(ctx0, GGML_TYPE_I32, N);
+  memcpy(embd->data, embd_inp.data(), N * ggml_element_size(embd));
+
+  struct ggml_tensor *inpL = ggml_get_rows(ctx0, model.wte_weight, embd);
+
+  for (int il = 0; il < n_layer; ++il) {
+
+    struct ggml_tensor *cur;
+
+    // a = self.ln_1(x)
+    {
+      cur = ggml_norm(ctx0, inpL);
```

Comment:
I think we have to use `ggml_rms_norm()` instead of `ggml_norm()`:

https://huggingface.co/mosaicml/mpt-7b/blob/main/norm.py#L27-L31

Reply:
the MPT-7B models are using `LPLayerNorm` https://huggingface.co/mosaicml/mpt-7b/blob/main/norm.py#L14

config: https://huggingface.co/mosaicml/mpt-7b/blob/main/config.json#L44